railties/lib/rails/application.rb

```diff
@@ -590,5 +590,46 @@ def build_request(env)
       def build_middleware
         config.app_middleware + super
       end
+
+      class NonSymbolAccessDeprecatedHash < Hash # :nodoc:
+        def initialize(value = nil)
+          if value.is_a?(Hash)
+            value.each_pair { |k, v| self[k] = v }
+          else
+            super
+          end
+        end
+
+        def []=(key, value)
+          if value.is_a?(Hash)
+            value = self.class.new(value)
+          end
+
+          super(key.to_sym, value)
+        end
+
+        def [](key)
+          super(convert_key(key))
+        end
+
+        def key?(key)
+          super(convert_key(key))
+        end
+
```

Comment:
Would `slice`, `extract` and `except` and the `!` versions with string keys work? How about `fetch`?

Reply:
They wouldn't work with the current state of the PR, but I can add similar implementation for those as well.

Reply:
Added support for these methods and a test that asserts they all work properly with string keys with deprecations.